cmd/api/auth/impersonation.go

```diff
@@ -0,0 +1,230 @@
+
+	// service account
+	if err == nil {
+		return &apiAuthzv1.ResourceAttributes{
+			Name:      sa,
+			Namespace: namespace,
+			Resource:  "serviceaccounts",
+			Verb:      "impersonate",
+		}, impersonatedUser, namespace
+
+	}
+
+	// user
+	return &apiAuthzv1.ResourceAttributes{
+		Name:     impersonatedUser,
+		Resource: "users",
+		Verb:     "impersonate",
+	}, impersonatedUser, ""
+
+}
+
+// parseGroups returns the SAR ResourceAttributes for impersonation and the groups from HTTP impersonation headers
+func parseGroups(headers http.Header) ([]*apiAuthzv1.ResourceAttributes, []string) {
+	groups := headers.Values(apiAuthnv1.ImpersonateGroupHeader)
+	resourceAtts := make([]*apiAuthzv1.ResourceAttributes, 0)
+	for _, group := range groups {
+		resourceAtts = append(resourceAtts, &apiAuthzv1.ResourceAttributes{
+			Name:     group,
+			Resource: "groups",
+			Verb:     "impersonate",
+		})
+	}
+	return resourceAtts, groups
+}
+
+// parseUID returns the SAR ResourceAttribute for impersonation and the UID from HTTP impersonation headers
+func parseUID(headers http.Header) (*apiAuthzv1.ResourceAttributes, string) {
+	uid := headers.Get(apiAuthnv1.ImpersonateUIDHeader)
+	if uid == "" {
+		return nil, ""
+	}
+	return &apiAuthzv1.ResourceAttributes{
+		Group:    apiAuthnv1.SchemeGroupVersion.Group,
+		Name:     uid,
+		Resource: "uids",
+		Verb:     "impersonate",
+	}, uid
+}
+
+// parseExtras returns the SAR ResourceAttributes for impersonation and the extras from HTTP impersonation headers
+func parseExtras(headers http.Header) ([]*apiAuthzv1.ResourceAttributes, map[string][]string) {
+	extras := make(map[string][]string)
+	resourceAtts := make([]*apiAuthzv1.ResourceAttributes, 0)
+	for headerKey, headerValues := range headers {
+		if strings.HasPrefix(headerKey, apiAuthnv1.ImpersonateUserExtraHeaderPrefix) {
+			encodedKey := strings.TrimPrefix(headerKey, apiAuthnv1.ImpersonateUserExtraHeaderPrefix)
+			key, err := url.PathUnescape(encodedKey)
+			if err != nil {
+				key = encodedKey
+			}
+			for _, value := range headerValues {
+				extras[key] = append(extras[key], value)
+				resourceAtts = append(resourceAtts, &apiAuthzv1.ResourceAttributes{
+					Group:       apiAuthnv1.SchemeGroupVersion.Group,
+					Name:        value,
+					Resource:    "userextras",
+					Subresource: key,
+					Verb:        "impersonate",
+				})
+			}
+		}
+	}
+	return resourceAtts, extras
+}
+
+func Impersonation(
+	sari clientAuthzv1.SubjectAccessReviewInterface,
+	cache *cache.Cache,
+	cacheExpirationAuthorized,
+	cacheExpirationUnauthorized time.Duration) gin.HandlerFunc {
+
+	if os.Getenv(impersonateFlag) != "true" {
+		return func(c *gin.Context) {
+			c.Next()
+		}
+	}
+
+	return func(c *gin.Context) {
+		imp, imperErr := newImpersonatedData(c)
+		if imperErr != nil && !errors.Is(imperErr, ErrNoImpersonationData) {
+			abort.Abort(c, fmt.Errorf("invalid impersonation data"), http.StatusBadRequest)
+			return
+		}
```

Comment:
I think we should fail on any other unexpected error., for example the error where it says that the impersonation header is required.

Reply:
I replaced the new created error for passing the error that comes from the parser